### PR TITLE
arch(burndown): drain positional-RPC indexing allowlist behind bound locals / row adapters (#1389)

### LINKED ISSUE #1389: arch(burndown): drain positional-RPC indexing allowlist behind safe_index / row adapters
## Burndown: migrate raw positional-RPC decoding behind `_row_adapters/` + `safe_index`

Follow-up to #1377. The durable GATE (`tests/_lint/test_no_raw_positional_rpc_indexing.py`) now forbids raw *chained* positional indexing of RPC payloads (`x[i][j]`) outside `rpc/` and `_row_adapters/`. The current offenders are baselined into that lint's `ALLOWLIST`. This issue tracks draining that allowlist — migrating each file behind `rpc/_safe_index.safe_index()` and/or a typed `_row_adapters/` view so genuine shape drift **raises** `UnknownRPCMethodError` (ADR-011) instead of silently degrading to empty/wrong data.

**The allowlist to drain** (each removal re-arms the gate for that file via the lint's `test_no_stale_allowlist_entries`):

- [ ] `_chat/wire.py` — `parse_citations` / `parse_single_citation` (`first[2][0]`, `cite[0][0]`, `first[4][-1]`). NOTE: every chained descent here is already guarded by an `isinstance`/`len` check where the absent case is *legitimate* (non-answer / non-citation records). Migration must preserve "no citations present ⇒ `[]`" and only raise on descents that are genuinely committed; treat with care.
- [ ] `_source/content.py` — `get_guide` / `get_fulltext` positional descents into `GET_SOURCE_GUIDE` / `GET_SOURCE`.
- [ ] `_research_task_parser.py` — residual `src[1][0]` / `src[1][1]` / `result[0][0]`; this module already uses `safe_index` for the task-id/info descents, so it is the most precedented next target.
- [ ] `_mind_maps_api.py` — `_new_artifact_id` (`create_response[0][0]`).
- [ ] `_note_service.py` — note-id / row-content descents (`item[1][0]`, `result[0][0]`).
- [ ] `_settings.py` — NOT caught by the gate (it walks `result[idx]` in a loop via `_extract_nested_value`, not chained literal subscripts), but listed in #1377 as an ad-hoc positional parser; converge it onto `safe_index` for the same drift-raises guarantee.
- [ ] `_artifacts.py` — `result[0][0]` shape probe + `params[2][2]` (request-side).
- [ ] `_artifact/listing.py` — `result[0][0]` shape probe.
- [ ] `_research.py` — `result[0][0]` / `src_data[0][0]` source descents.
- [ ] `_types/artifacts.py`, `_types/notebooks.py`, `_types/notes.py`, `_types/sharing.py` — dataclass `from_api_response`/`from_*` positional parsers.

**Shape-drift semantics (ADR-011):** for each site, distinguish "field legitimately absent/empty ⇒ return the empty/None default" from "structure drifted ⇒ raise via `safe_index`". The offenders are shape-discriminating parsers (multiple wire shapes coexist), so a blanket swallow→raise is unsafe — migrate per-site with the legitimate-absent cases preserved.

**Dead code to remove separately:** `Note.from_api_response` appears dead — note parsing flows through `NoteRow` directly. Confirm no live callers and delete in its own change.

Relates to ADR-011, #1344.

### PR DESCRIPTION
**Pull request by teng-lin**
## What

Drains the `tests/_lint/test_no_raw_positional_rpc_indexing.py` `ALLOWLIST` to **empty**: every feature file the chained-positional-subscript gate flagged (`x[i][j]` integer-literal descent outside `rpc/` / `_row_adapters/`) is migrated, so the gate now re-protects the whole feature tree.

Migrated files: `_artifact/listing.py`, `_artifacts.py`, `_chat/wire.py`, `_mind_maps_api.py`, `_note_service.py`, `_research.py`, `_research_task_parser.py`, `_source/content.py`, `_types/{artifacts,notebooks,notes,sharing}.py`.

Also removes the dead `Note.from_api_response`, syncs `uv.lock` ruff `0.15.14 → 0.15.15` to match the committed `pyproject` pin, and adds a `parse_single_citation` absent-chunk-id test.

## Why

Per **ADR-011**, hand-rolled chained positional descent into batchexecute payloads re-scatters position knowledge and routinely *swallows* Google-side shape drift to an empty/wrong value behind `try/except (IndexError, TypeError)`. The durable gate (#1377) baselined the offenders; this is the #1389 burndown that drains that baseline.

## How

Each migrated descent **binds the already-guarded inner list to a named local** so each leaf read is a single-level index, keeping the existing `isinstance`/`len` guards verbatim. This is the migration shape the gate's own docstring sanctions ("bound the already-guarded inner list to a named local so each leaf read is a single-level index").

Crucially, **no swallow→raise flip was forced** on any optional-field site — that would be unsafe (ADR-011: legitimate-absent ⇒ default, only genuine drift ⇒ raise). So:

- **`_chat/wire.py` (most delicate):** `parse_citations` keeps "no citations present ⇒ `[]`" and its `try/except → []` guard untouched. `parse_single_citation` / `_extract_chunk_with_parseable` move the optional chunk-id, server-conversation-id, and answer-flag leaf reads to bound locals; an absent block still yields `None`/`False`/kept-citation.
- **`_source/content.py`:** absent summary/keywords/html-rendition/text blocks still default (covered by `test_get_guide_shape_variants_return_stable_defaults` + the missing-rendition/empty-text tests).
- **shape probes** (`_artifacts.py`, `_artifact/listing.py`, `_mind_maps_api.py`, `_research.py`, `_research_task_parser._unwrap_poll_result`): the wrap-vs-flat envelope detection is preserved exactly; a degenerate response still returns the same default/None.
- **`_note_service.py`:** the `if not note_id: raise RPCError(...)` guard means a drifted `CREATE_NOTE` shape still surfaces an error rather than fabricating an empty-id note — a site where drift correctly **raises**.
- **`_types/*` dataclass parsers:** optional timestamp/owner-flag/is-public blocks still default to `None`/`True`/`False` on absence.

`Note.from_api_response` is removed as dead code (notes decode via `NoteRow`; `Note(...)` is built in `NotesAPI._parse_note`; no production callers — only `test_types.py` exercised it). Recorded in `scripts/api-compat-allowlist.json` under both path-views; the audit passes (`OK: public API is compatible with v0.6.0`).

The `uv.lock` ruff bump is a **legitimate sync** — `main`'s `pyproject.toml` already pins `ruff==0.15.15` while `main`'s lock recorded `0.15.14`, so `uv lock --check` fails on `main`; this corrects that drift (only the ruff package version + its hashes change).

`_artifacts.py` / `_research.py` sit exactly at their ADR-0008 module-size ceilings, so the migration comments were kept terse to stay net-neutral and respect the ratchet.

## Verification

- `uv run pytest tests/unit tests/_lint` → **7409 passed, 106 skipped, 1 xfailed**
- `uv run pytest tests/integration` → **707 passed, 2 skipped**
- `uv run pytest tests/_lint/test_no_raw_positional_rpc_indexing.py` → 7 passed (ALLOWLIST empty, gate re-armed)
- `uv run pytest tests/_lint/test_module_size_ratchet.py` → green (both files at ceiling)
- `uv run python scripts/audit_public_api_compat.py` → `OK` (only the two intentional `Note.from_api_response` removals are new)
- `uv lock --check` → clean
- `uv run mypy src/notebooklm --ignore-missing-imports` → `Success: no issues found in 195 source files`
- `uv run ruff format` + `uv run ruff check` → clean

## Residual

References #1389 (does not auto-close). The lint ALLOWLIST that #1389 mechanically tracks is fully drained, but one #1389 checklist item is deliberately deferred:

- **`_settings.py`** — listed on the #1389 checklist but **not** caught by the AST gate (it path-walks `result[idx]` in a loop, not literal chained subscripts). Its optional current-language slot must return `None` when legitimately absent, so a blanket `safe_index` conversion would regress the optional-language contract (the exact "blanket swallow→raise is unsafe" hazard #1389 warns about). Left on its bounded `try/except → None` form and tracked as follow-up **#1395**.

Maintainer can close #1389 once the `_settings.py` deferral is accepted.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved handling of malformed or missing API response blocks to reduce runtime errors and increase stability.

* **Chores**
  * Removed a deprecated legacy note-parsing entry.
  * Refactored internal response parsing to be safer and clearer.

* **Tests**
  * Added unit coverage for citation parsing edge cases and tightened a lint allowlist to enforce safer indexing.

### CODE REVIEW

**Comment by teng-lin** [comment]
@claude review

File: tests/unit/test_types.py [hunk 1]
```diff
@@ -170,7 +170,6 @@ def test_representative_public_dataclasses_pickle_round_trip():
         ConnectionLimits,
         ConversationTurn,
         GenerationStatus,
-        Note,
         Notebook,
```
[the remark this answers is not in the document]
Comment (teng-lin) [reply tests/unit/test_types.py:173]: Good eye, and the intent here (verify `Note` stays importable from `notebooklm.types` and pickles) is preserved — but applying the suggestion would actually break `ruff check`.

`Note` is already imported from `notebooklm.types` at **module scope** (test_types.py line 17), so the `Note(...)` instantiation at line 222 and the pickle round-trip both resolve against that module-level binding. Re-adding `Note` to the function-local import block would be a redefinition of the module-level name and trips ruff `F811` (redefinition / shadowing) — which is exactly the error I hit when I tried that fix, so I reverted it.

So the importability+pickle coverage you're asking for is intact via the module-level import; the function-local re-import was redundant and was correctly dropped when `Note.from_api_response` (the only reason that block needed `Note` locally) was deleted. Leaving as-is to keep `ruff check` green.

**Comment by teng-lin** [comment]
Thanks for the thorough review (claude / cubic / coderabbit / gemini — all green, no bugs found). Addressing the three non-blocking nits:

1. **`_source/content.py` `else []` vs `else None` inconsistency** — Fixed in `541ba1b`: aligned both `summary_block` / `keyword_block` to `else None` to match every other migrated site in the PR. Behavior-preserving (the `if block:` guard is falsy for both), still covered by `test_get_guide_shape_variants_return_stable_defaults`.

2. **`test_types.py` local `Note` import consistency** (also raised by gemini) — Intentionally left as-is. `Note` is imported at module scope (line 17, from `notebooklm.types`), so the `Note(...)` instance at line 222 and the pickle round-trip already resolve against it — the importability+pickle coverage is intact. Re-adding `Note` to the function-local import block would redefine the module-level name and trip ruff `F811`, so it stays dropped.

3. **Comment verbosity at migration sites** — Trimmed the per-site comments to 1–3 lines (the `_artifacts.py` / `_research.py` sites were already tightened to stay under their ADR-0008 module-size ceilings). Kept a one-line WHY at each site since the bound-local pattern's intent (lint-driven, ADR-011 legitimate-absent vs drift) isn't obvious from variable names alone.
